File: backends/credhub/src/main/java/org/cloudfoundry/credhub/credentials/CredentialsController.java
package org.cloudfoundry.credhub.credentials;

import java.io.ByteArrayInputStream;
import java.io.IOException;
import java.io.InputStream;

import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.http.HttpStatus;
import org.springframework.http.MediaType;
import org.springframework.web.bind.annotation.GetMapping;
import org.springframework.web.bind.annotation.PathVariable;
import org.springframework.web.bind.annotation.RequestBody;
import org.springframework.web.bind.annotation.RequestMapping;
import org.springframework.web.bind.annotation.RequestMethod;
import org.springframework.web.bind.annotation.RequestParam;
import org.springframework.web.bind.annotation.ResponseStatus;
import org.springframework.web.bind.annotation.RestController;

import com.google.common.io.ByteStreams;
import org.apache.commons.lang3.StringUtils;
import org.cloudfoundry.credhub.ErrorMessages;
import org.cloudfoundry.credhub.audit.CEFAuditRecord;
import org.cloudfoundry.credhub.audit.RequestDetails;
import org.cloudfoundry.credhub.audit.entities.DeleteCredential;
import org.cloudfoundry.credhub.audit.entities.FindCredential;
import org.cloudfoundry.credhub.audit.entities.GetCredential;
import org.cloudfoundry.credhub.audit.entities.SetCredential;
import org.cloudfoundry.credhub.exceptions.InvalidQueryParameterException;
import org.cloudfoundry.credhub.generate.CredentialsHandler;
import org.cloudfoundry.credhub.generate.LegacyGenerationHandler;
import org.cloudfoundry.credhub.generate.SetHandler;
import org.cloudfoundry.credhub.requests.BaseCredentialSetRequest;
import org.cloudfoundry.credhub.services.PermissionedCredentialService;
import org.cloudfoundry.credhub.views.CredentialView;
import org.cloudfoundry.credhub.views.DataResponse;
import org.cloudfoundry.credhub.views.FindCredentialResults;

@RestController
@RequestMapping(
  path = CredentialsController.ENDPOINT,
  produces = MediaType.APPLICATION_JSON_UTF8_VALUE
)
public class CredentialsController {

  public static final String ENDPOINT = "/api/v1/data";

  private final PermissionedCredentialService permissionedCredentialService;
  private final SetHandler setHandler;
  private final CredentialsHandler credentialsHandler;
  private final LegacyGenerationHandler legacyGenerationHandler;
  private final CEFAuditRecord auditRecord;

  @Autowired
  public CredentialsController(
    final PermissionedCredentialService permissionedCredentialService,
    final CredentialsHandler credentialsHandler,
    final SetHandler setHandler,
    final LegacyGenerationHandler legacyGenerationHandler,
    final CEFAuditRecord auditRecord
  ) {
    super();
    this.permissionedCredentialService = permissionedCredentialService;
    this.credentialsHandler = credentialsHandler;
    this.setHandler = setHandler;
    this.legacyGenerationHandler = legacyGenerationHandler;
    this.auditRecord = auditRecord;
  }

  @RequestMapping(path = "", method = RequestMethod.POST)
  @ResponseStatus(HttpStatus.OK)
  public synchronized CredentialView generate(final InputStream inputStream) throws IOException {
    final InputStream requestInputStream = new ByteArrayInputStream(ByteStreams.toByteArray(inputStream));
    return legacyGenerationHandler.auditedHandlePostRequest(requestInputStream);
  }

  @RequestMapping(path = "", method = RequestMethod.PUT)
  @ResponseStatus(HttpStatus.OK)
  public synchronized CredentialView set(@RequestBody final BaseCredentialSetRequest requestBody) {
    requestBody.validate();
    return auditedHandlePutRequest(requestBody);
  }

  @RequestMapping(path = "", method = RequestMethod.DELETE)
  @ResponseStatus(HttpStatus.NO_CONTENT)
  public void delete(@RequestParam("name") final String credentialName) {
    if (StringUtils.isEmpty(credentialName)) {
      throw new InvalidQueryParameterException(ErrorMessages.MISSING_QUERY_PARAMETER, "name");
    }

    final String credentialNameWithPrependedSlash = StringUtils.prependIfMissing(credentialName, "/");

    final RequestDetails requestDetails = new DeleteCredential(credentialNameWithPrependedSlash);
    auditRecord.setRequestDetails(requestDetails);

    credentialsHandler.deleteCredential(credentialNameWithPrependedSlash);
  }

  @RequestMapping(path = "/{id}", method = RequestMethod.GET)
  @ResponseStatus(HttpStatus.OK)
  public CredentialView findById(@PathVariable final String id) {
    return credentialsHandler.getCredentialVersionByUUID(id);
  }

  @GetMapping(path = "")
  @ResponseStatus(HttpStatus.OK)
  public DataResponse getByName(
    @RequestParam("name") final String credentialName,
    @RequestParam(value = "versions", required = false) final Integer numberOfVersions,
    @RequestParam(value = "current", required = false, defaultValue = "false") final boolean current
  ) {
    if (StringUtils.isEmpty(credentialName)) {
      throw new InvalidQueryParameterException(ErrorMessages.MISSING_QUERY_PARAMETER, "name");
    }

    if (current && numberOfVersions != null) {
      throw new InvalidQueryParameterException(ErrorMessages.CANT_USE_VERSIONS_AND_CURRENT, "name");
    }

    final String credentialNameWithPrependedSlash = StringUtils.prependIfMissing(credentialName, "/");

    auditRecord.setRequestDetails(new GetCredential(credentialName, numberOfVersions, current));

    if (current) {
      return credentialsHandler.getCurrentCredentialVersions(credentialNameWithPrependedSlash);
    } else {
      return credentialsHandler.getNCredentialVersions(credentialNameWithPrependedSlash, numberOfVersions);
    }
  }

  @RequestMapping(path = "", params = "path", method = RequestMethod.GET)
  @ResponseStatus(HttpStatus.OK)
  public FindCredentialResults findByPath(
    @RequestParam("path") final String path,
    @RequestParam(value = "expires-within-days", required = false, defaultValue = "") final String expiresWithinDays
  ) {
    final FindCredential findCredential = new FindCredential();
    findCredential.setPath(path);
    findCredential.setExpiresWithinDays(expiresWithinDays);
    auditRecord.setRequestDetails(findCredential);

    return new FindCredentialResults(permissionedCredentialService.findStartingWithPath(path, expiresWithinDays));
  }

  @RequestMapping(path = "", params = "name-like", method = RequestMethod.GET)
  @ResponseStatus(HttpStatus.OK)
  public FindCredentialResults findByNameLike(
    @RequestParam("name-like") final String nameLike,
    @RequestParam(value = "expires-within-days", required = false, defaultValue = "") final String expiresWithinDays
  ) {
    final FindCredential findCredential = new FindCredential();
    findCredential.setNameLike(nameLike);
    findCredential.setExpiresWithinDays(expiresWithinDays);
    auditRecord.setRequestDetails(findCredential);

    return new FindCredentialResults(permissionedCredentialService.findContainingName(nameLike, expiresWithinDays));
  }

  private CredentialView auditedHandlePutRequest(@RequestBody final BaseCredentialSetRequest requestBody) {
    auditRecord.setRequestDetails(new SetCredential(requestBody.getName(), requestBody.getType()));
    return setHandler.handle(requestBody);
  }
}

File: backends/credhub/src/main/kotlin/org/cloudfoundry/credhub/credentials/CredentialsController.kt
package org.cloudfoundry.credhub.credentials

import com.fasterxml.jackson.databind.JsonNode
import com.fasterxml.jackson.databind.ObjectMapper
import java.io.IOException
import org.apache.commons.lang3.StringUtils
import org.cloudfoundry.credhub.ErrorMessages
import org.cloudfoundry.credhub.audit.CEFAuditRecord
import org.cloudfoundry.credhub.audit.entities.DeleteCredential
import org.cloudfoundry.credhub.audit.entities.FindCredential
import org.cloudfoundry.credhub.audit.entities.GenerateCredential
import org.cloudfoundry.credhub.audit.entities.GetCredential
import org.cloudfoundry.credhub.audit.entities.RegenerateCredential
import org.cloudfoundry.credhub.audit.entities.SetCredential
import org.cloudfoundry.credhub.exceptions.InvalidQueryParameterException
import org.cloudfoundry.credhub.regenerate.RegenerateHandler
import org.cloudfoundry.credhub.requests.BaseCredentialGenerateRequest
import org.cloudfoundry.credhub.requests.BaseCredentialSetRequest
import org.cloudfoundry.credhub.requests.CredentialRegenerateRequest
import org.cloudfoundry.credhub.views.CredentialView
import org.cloudfoundry.credhub.views.DataResponse
import org.cloudfoundry.credhub.views.FindCredentialResults
import org.springframework.http.HttpStatus
import org.springframework.http.MediaType
import org.springframework.web.bind.annotation.PathVariable
import org.springframework.web.bind.annotation.RequestBody
import org.springframework.web.bind.annotation.RequestMapping
import org.springframework.web.bind.annotation.RequestMethod.DELETE
import org.springframework.web.bind.annotation.RequestMethod.GET
import org.springframework.web.bind.annotation.RequestMethod.POST
import org.springframework.web.bind.annotation.RequestMethod.PUT
import org.springframework.web.bind.annotation.RequestParam
import org.springframework.web.bind.annotation.ResponseStatus
import org.springframework.web.bind.annotation.RestController

@RestController
@RequestMapping(
    path = [CredentialsController.ENDPOINT],
    produces = [MediaType.APPLICATION_JSON_UTF8_VALUE]
)
class CredentialsController(
    private val credentialsHandler: CredentialsHandler,
    private val auditRecord: CEFAuditRecord,
    val regenerateHandler: RegenerateHandler,
    private val objectMapper: ObjectMapper
) {
    companion object {
        const val ENDPOINT = "/api/v1/data"
    }

    @RequestMapping(method = [GET], path = ["/{id}"])
    @ResponseStatus(HttpStatus.OK)
    fun getById(@PathVariable id: String): CredentialView {
        return credentialsHandler.getCredentialVersionByUUID(id)
    }

    @RequestMapping(method = [GET], path = [""])
    @ResponseStatus(HttpStatus.OK)
    fun getByName(
        @RequestParam("name") credentialName: String,
        @RequestParam(value = "versions", required = false) numberOfVersions: Int?,
        @RequestParam(value = "current", required = false, defaultValue = "false") current: Boolean
    ): DataResponse {
        if (StringUtils.isEmpty(credentialName)) {
            throw InvalidQueryParameterException(ErrorMessages.MISSING_QUERY_PARAMETER, "name")
        }

        if (current && numberOfVersions != null) {
            throw InvalidQueryParameterException(ErrorMessages.CANT_USE_VERSIONS_AND_CURRENT, "name")
        }

        val credentialNameWithPrependedSlash = StringUtils.prependIfMissing(credentialName, "/")

        auditRecord.requestDetails = GetCredential(credentialName, numberOfVersions, current)

        return if (current) {
            credentialsHandler.getCurrentCredentialVersions(credentialNameWithPrependedSlash)
        } else {
            credentialsHandler.getNCredentialVersions(credentialNameWithPrependedSlash, numberOfVersions)
        }
    }

    @RequestMapping(method = [GET], path = [""], params = ["path"])
    @ResponseStatus(HttpStatus.OK)
    fun findByPath(
        @RequestParam("path") path: String,
        @RequestParam(value = "expires-within-days", required = false, defaultValue = "") expiresWithinDays: String
    ): FindCredentialResults {
        val findCredential = FindCredential()
        findCredential.path = path
        findCredential.expiresWithinDays = expiresWithinDays
        auditRecord.requestDetails = findCredential

        return FindCredentialResults(credentialsHandler.findStartingWithPath(path, expiresWithinDays))
    }

    @RequestMapping(method = [GET], path = [""], params = ["name-like"])
    @ResponseStatus(HttpStatus.OK)
    fun findByNameLike(
        @RequestParam("name-like") nameLike: String,
        @RequestParam(value = "expires-within-days", required = false, defaultValue = "") expiresWithinDays: String
    ): FindCredentialResults {
        val findCredential = FindCredential()
        findCredential.nameLike = nameLike
        findCredential.expiresWithinDays = expiresWithinDays
        auditRecord.requestDetails = findCredential

        return FindCredentialResults(credentialsHandler.findContainingName(nameLike, expiresWithinDays))
    }

    @Synchronized
    @RequestMapping(method = [POST], path = [""])
    @ResponseStatus(HttpStatus.OK)
    @Throws(IOException::class)
    fun generate(@RequestBody request: JsonNode): CredentialView {
        val isRegenerate = request["regenerate"]?.asBoolean() ?: false
        try {
            if (isRegenerate) {
                val regenerateRequest = objectMapper.readValue(request.toString(), CredentialRegenerateRequest::class.java)
                regenerateRequest.validate()
                val regenerateCredential = RegenerateCredential()
                regenerateCredential.name = regenerateRequest.name
                auditRecord.requestDetails = regenerateCredential

                return regenerateHandler.handleRegenerate(regenerateRequest.name)
            } else {
                val requestString = request.toString()
                val generateRequest = objectMapper.readValue(requestString, BaseCredentialGenerateRequest::class.java)
                generateRequest.validate()
                val generateCredential = GenerateCredential()
                generateCredential.name = generateRequest.name
                generateCredential.type = generateRequest.type
                auditRecord.requestDetails = generateCredential

                return credentialsHandler.generateCredential(generateRequest)
            }
        } catch (e: IOException) {
            throw RuntimeException(e)
        }
    }

    @RequestMapping(method = [PUT], path = [""])
    @ResponseStatus(HttpStatus.OK)
    @Synchronized
    fun set(@RequestBody requestBody: BaseCredentialSetRequest<*>): CredentialView {
        requestBody.validate()
        auditRecord.requestDetails = SetCredential(requestBody.name, requestBody.type)

        return credentialsHandler.setCredential(requestBody)
    }

    @RequestMapping(method = [DELETE], path = [""])
    @ResponseStatus(HttpStatus.NO_CONTENT)
    fun delete(@RequestParam("name") credentialName: String) {
        if (StringUtils.isEmpty(credentialName)) {
            throw InvalidQueryParameterException(ErrorMessages.MISSING_QUERY_PARAMETER, "name")
        }

        val credentialNameWithPrependedSlash = StringUtils.prependIfMissing(credentialName, "/")

        val requestDetails = DeleteCredential(credentialNameWithPrependedSlash)
        auditRecord.requestDetails = requestDetails

        credentialsHandler.deleteCredential(credentialNameWithPrependedSlash)
    }
}
